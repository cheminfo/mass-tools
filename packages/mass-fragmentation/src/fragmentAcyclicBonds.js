import { MF } from 'mf-parser';
import { getMF, getHoseCodesForAtoms } from 'openchemlib-utils';

/**
 * The function performs the fragmentation of all single linear bonds
 * @param {import('openchemlib').Molecule} molecule - The OCL molecule to be fragmented
 * @param {object} [options={}]
 * @param {boolean} [options.calculateHoseCodes=false] - calculating hose code for bonds is quite time consuming
 * @returns Results fragmentation of acyclic bonds
 */

export function fragmentAcyclicBonds(molecule, options = {}) {
  const { Molecule } = molecule.getOCL();
  const { calculateHoseCodes } = options;
  let atoms = [];
  // Prepare object with lenght equal to number of atoms
  for (let i = 0; i < molecule.getAllAtoms(); i++) {
    let atom = {};
    atoms.push(atom);
    atom.i = i;
    atom.links = [];
  }
  let bonds = [];
  for (let i = 0; i < molecule.getAllBonds(); i++) {
    let bond = {};
    // get informations of bonds
    bond.index = i;
    bond.order = molecule.getBondOrder(i); // dative, single , double, triple
    bond.atom1 = molecule.getBondAtom(0, i); // atom 1 index
    bond.atom2 = molecule.getBondAtom(1, i); // atom 2 index

    bond.type = molecule.getBondType(i); // cBondTypeSingle,cBondTypeDouble,cBondTypeTriple,cBondTypeDelocalized
    bond.isAromatic = molecule.isAromaticBond(i);
    bond.isRingBond = molecule.isRingBond(i);

    // Mapping of bonds to be fragmented, only if they are single bond not aromatic and cyclic the mapping occurs
    if (
      bond.isAromatic ||
      bond.type > 1 ||
      bond.isRingBond ||
      bond.order !== 1
    ) {
      continue;
    } else {
      bond.selected = true;
      atoms[bond.atom1].links.push(bond.atom2);
      atoms[bond.atom2].links.push(bond.atom1);
    }
    bonds.push(bond);
  }

  let brokenMolecule = {};
  let fragmentMap = [];
  let nbFragments = [];
  let results = [
    {
      idCode: molecule.getIDCode(),
      mfInfo: new MF(getMF(molecule).mf).getInfo(),
      fragmentType: 'molecular ion',
    },
  ];

  for (let bond of bonds) {
    if (bond.selected) {
      // if bond.selected is true (line 46) the molecule will be fragmented
      brokenMolecule[bond.index] = molecule.getCompactCopy(); // get a copy of the molecule
      brokenMolecule[bond.index].setAtomCustomLabel(bond.atom1, '*');
      brokenMolecule[bond.index].setAtomCustomLabel(bond.atom2, '*');
      brokenMolecule[bond.index].markBondForDeletion(bond.index); //mark bond to be deleted
      // the function returns an array of map

      brokenMolecule[bond.index].deleteMarkedAtomsAndBonds(); // delete marked bonds
    }
    nbFragments = brokenMolecule[bond.index].getFragmentNumbers(fragmentMap);
    // only if there are 2 fragments code can continue
    if (nbFragments === 2) {
      for (let i = 0; i < nbFragments; i++) {
        const result = {};
        if (calculateHoseCodes) {
          result.hoses = getHoseCodesForAtoms(molecule, [
            bond.atom1,
            bond.atom2,
          ]);
        }

        result.atomMap = [];

        // assign fragment id to index of for loop
        let includeAtom = fragmentMap.map((id) => {
          return id === i;
        });

        let fragment = new Molecule(100, 100);

        let atomMap = [];

        brokenMolecule[bond.index].copyMoleculeByAtoms(
          fragment,
          includeAtom,
          false,
          atomMap,
        );

        for (let j = 0; j < atomMap.length; j++) {
          if (fragment.getAtomCustomLabel(atomMap[j]) === '*') {
            result.atomMap.push(j);
            if (atoms[j].links.length > 0) {
              fragment.addBond(atomMap[j], fragment.addAtom(154));
            }
          }
        }
        fragment.removeAtomCustomLabels();
        fragment.setFragment(false);
        result.idCode = fragment.getIDCode();
        result.cleavedBonds = [
          {
            index: bond.index,
            order: bond.order,
            atom1: bond.atom1,
            atom2: bond.atom2,
          },
        ];
        result.mfInfo = new MF(
          getMF(fragment).mf.replace(/R[1-9]?/, ''),
        ).getInfo();
        result.fragmentType = 'acyclic';

        results.push(result);
      }
    }
  }
  // sort result in order fragment 1-2; 3-4; ...
  results = results.sort((a, b) => {
    return a.mfInfo.monoisotopicMass - b.mfInfo.monoisotopicMass;
  });

  return results;
}
