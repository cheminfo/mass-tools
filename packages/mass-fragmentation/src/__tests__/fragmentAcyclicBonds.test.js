import OCL from 'openchemlib';

import { fragmentAcyclicBonds } from '../fragmentAcyclicBonds.js';

const { Molecule } = OCL;

describe('fragmentAcyclicBonds', () => {
  it('CCCC', () => {
    const molecule = Molecule.fromSmiles('CCCC'); // for some reason tostrictEqual has problems with: 'eMHAIhNFhF`QR\\Ji\\Jh'
    const result = fragmentAcyclicBonds(molecule);

    expect(result).toMatchSnapshot();
  });
  it('c3ccc(CCCC2CCC1CCCCC1C2)cc3 - options false', () => {
    const molecule = Molecule.fromSmiles('c3ccc(CCCC2CCC1CCCCC1C2)cc3');
    const result = fragmentAcyclicBonds(molecule, {
      calculateHoseCodes: false,
    });

    expect(result).toMatchSnapshot();
  });
  it('c3ccc(CCCC2CCC1CCCCC1C2)cc3 - options true', () => {
    const molecule = Molecule.fromSmiles('c3ccc(CCCC2CCC1CCCCC1C2)cc3');
    const result = fragmentAcyclicBonds(molecule, { calculateHoseCodes: true });

    expect(result).toMatchSnapshot();
  });
});
