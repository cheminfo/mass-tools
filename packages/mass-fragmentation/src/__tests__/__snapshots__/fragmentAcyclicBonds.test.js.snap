// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fragmentAcyclicBonds CCCC 1`] = `
[
  {
    "atomMap": [
      0,
    ],
    "cleavedBonds": [
      {
        "atom1": 0,
        "atom2": 1,
        "index": 0,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "eFBAYc@@",
    "mfInfo": {
      "atoms": {
        "C": 1,
        "H": 3,
      },
      "charge": 0,
      "mass": 15.03455815890258,
      "mf": "CH3",
      "monoisotopicMass": 15.02347509669,
      "unsaturation": 0.5,
    },
  },
  {
    "atomMap": [
      3,
    ],
    "cleavedBonds": [
      {
        "atom1": 2,
        "atom2": 3,
        "index": 2,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "eFBAYc@@",
    "mfInfo": {
      "atoms": {
        "C": 1,
        "H": 3,
      },
      "charge": 0,
      "mass": 15.03455815890258,
      "mf": "CH3",
      "monoisotopicMass": 15.02347509669,
      "unsaturation": 0.5,
    },
  },
  {
    "atomMap": [
      1,
    ],
    "cleavedBonds": [
      {
        "atom1": 1,
        "atom2": 2,
        "index": 1,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "eMBAYRZ@",
    "mfInfo": {
      "atoms": {
        "C": 2,
        "H": 5,
      },
      "charge": 0,
      "mass": 29.061175563749384,
      "mf": "C2H5",
      "monoisotopicMass": 29.03912516115,
      "unsaturation": 0.5,
    },
  },
  {
    "atomMap": [
      2,
    ],
    "cleavedBonds": [
      {
        "atom1": 1,
        "atom2": 2,
        "index": 1,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "eMBAYRZ@",
    "mfInfo": {
      "atoms": {
        "C": 2,
        "H": 5,
      },
      "charge": 0,
      "mass": 29.061175563749384,
      "mf": "C2H5",
      "monoisotopicMass": 29.03912516115,
      "unsaturation": 0.5,
    },
  },
  {
    "atomMap": [
      1,
    ],
    "cleavedBonds": [
      {
        "atom1": 0,
        "atom2": 1,
        "index": 0,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "gC\`H@liKT@@",
    "mfInfo": {
      "atoms": {
        "C": 3,
        "H": 7,
      },
      "charge": 0,
      "mass": 43.08779296859618,
      "mf": "C3H7",
      "monoisotopicMass": 43.05477522561,
      "unsaturation": 0.5,
    },
  },
  {
    "atomMap": [
      2,
    ],
    "cleavedBonds": [
      {
        "atom1": 2,
        "atom2": 3,
        "index": 2,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "gC\`H@liKT@@",
    "mfInfo": {
      "atoms": {
        "C": 3,
        "H": 7,
      },
      "charge": 0,
      "mass": 43.08779296859618,
      "mf": "C3H7",
      "monoisotopicMass": 43.05477522561,
      "unsaturation": 0.5,
    },
  },
  {
    "fragmentType": "molecular ion",
    "idCode": "gC\`@Dij@@",
    "mfInfo": {
      "atoms": {
        "C": 4,
        "H": 10,
      },
      "charge": 0,
      "mass": 58.12235112749877,
      "mf": "C4H10",
      "monoisotopicMass": 58.0782503223,
      "unsaturation": 0,
    },
  },
]
`;

exports[`fragmentAcyclicBonds c3ccc(CCCC2CCC1CCCCC1C2)cc3 - options false 1`] = `
[
  {
    "atomMap": [
      3,
    ],
    "cleavedBonds": [
      {
        "atom1": 3,
        "atom2": 4,
        "index": 3,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "gOpH@liLkW@@@@",
    "mfInfo": {
      "atoms": {
        "C": 6,
        "H": 5,
      },
      "charge": 0,
      "mass": 77.10411915069038,
      "mf": "C6H5",
      "monoisotopicMass": 77.03912516115,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      4,
    ],
    "cleavedBonds": [
      {
        "atom1": 4,
        "atom2": 5,
        "index": 4,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "daD@\`@VTeeVz\`@@@",
    "mfInfo": {
      "atoms": {
        "C": 7,
        "H": 7,
      },
      "charge": 0,
      "mass": 91.13073655553718,
      "mf": "C7H7",
      "monoisotopicMass": 91.05477522561,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      5,
    ],
    "cleavedBonds": [
      {
        "atom1": 5,
        "atom2": 6,
        "index": 5,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "did@\`@VTeYWaj@@@@",
    "mfInfo": {
      "atoms": {
        "C": 8,
        "H": 9,
      },
      "charge": 0,
      "mass": 105.15735396038399,
      "mf": "C8H9",
      "monoisotopicMass": 105.07042529007,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      6,
    ],
    "cleavedBonds": [
      {
        "atom1": 6,
        "atom2": 7,
        "index": 6,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "deT@\`@VTeVUFZh@@@@",
    "mfInfo": {
      "atoms": {
        "C": 9,
        "H": 11,
      },
      "charge": 0,
      "mass": 119.18397136523078,
      "mf": "C9H11",
      "monoisotopicMass": 119.08607535453,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      7,
    ],
    "cleavedBonds": [
      {
        "atom1": 6,
        "atom2": 7,
        "index": 6,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "dmL@\`@VTfUmZZUjjjj@@",
    "mfInfo": {
      "atoms": {
        "C": 10,
        "H": 17,
      },
      "charge": 0,
      "mass": 137.2423517863007,
      "mf": "C10H17",
      "monoisotopicMass": 137.13302554791,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      6,
    ],
    "cleavedBonds": [
      {
        "atom1": 5,
        "atom2": 6,
        "index": 5,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "dcl@\`@VTee[UnWZjjjh@@",
    "mfInfo": {
      "atoms": {
        "C": 11,
        "H": 19,
      },
      "charge": 0,
      "mass": 151.2689691911475,
      "mf": "C11H19",
      "monoisotopicMass": 151.14867561237,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      5,
    ],
    "cleavedBonds": [
      {
        "atom1": 4,
        "atom2": 5,
        "index": 4,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "dk\\@\`@VTeYVux]Njjjj\`@",
    "mfInfo": {
      "atoms": {
        "C": 12,
        "H": 21,
      },
      "charge": 0,
      "mass": 165.2955865959943,
      "mf": "C12H21",
      "monoisotopicMass": 165.16432567683,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      4,
    ],
    "cleavedBonds": [
      {
        "atom1": 3,
        "atom2": 4,
        "index": 3,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "idCode": "dg|@\`@VTeVUmQd{jjjjj@@",
    "mfInfo": {
      "atoms": {
        "C": 13,
        "H": 23,
      },
      "charge": 0,
      "mass": 179.3222040008411,
      "mf": "C13H23",
      "monoisotopicMass": 179.17997574129,
      "unsaturation": 2.5,
    },
  },
  {
    "fragmentType": "molecular ion",
    "idCode": "flu@@@DjU_YVuirshiBBJjjjJ@@",
    "mfInfo": {
      "atoms": {
        "C": 19,
        "H": 28,
      },
      "charge": 0,
      "mass": 256.4263231515315,
      "mf": "C19H28",
      "monoisotopicMass": 256.21910090244,
      "unsaturation": 6,
    },
  },
]
`;

exports[`fragmentAcyclicBonds c3ccc(CCCC2CCC1CCCCC1C2)cc3 - options true 1`] = `
[
  {
    "atomMap": [
      3,
    ],
    "cleavedBonds": [
      {
        "atom1": 3,
        "atom2": 4,
        "index": 3,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gJP@DjXHCaB\`eAzSOtRApjhxUP",
      "dax@@DjUXHH@xPJBJCqQbQdEgzH\`NEUCaU@",
      "deT@@DjUWXXHH\`@xPJBJCwQbQdEePt]gzH\`NEUCaU@",
      "dcL@@DjUWya\`\`bh@NDB\`b\`}LXdYAYTMGXLSGzH\`NEUCaU@",
    ],
    "idCode": "gOpH@liLkW@@@@",
    "mfInfo": {
      "atoms": {
        "C": 6,
        "H": 5,
      },
      "charge": 0,
      "mass": 77.10411915069038,
      "mf": "C6H5",
      "monoisotopicMass": 77.03912516115,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      4,
    ],
    "cleavedBonds": [
      {
        "atom1": 4,
        "atom2": 5,
        "index": 4,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
      "gNp@DiYjB@xPhIP^prYlQHGBjcaU@",
      "dmT@@DiYUj\`\`h@xPJBJCsAdYfEEPt_hb@xUTNET",
      "dg\\@@DiYUgQvhHJJ@Ca@hHhOJFQfXTUCQtCDxn_hb@xUTNET",
    ],
    "idCode": "daD@\`@VTeeVz\`@@@",
    "mfInfo": {
      "atoms": {
        "C": 7,
        "H": 7,
      },
      "charge": 0,
      "mass": 91.13073655553718,
      "mf": "C7H7",
      "monoisotopicMass": 91.05477522561,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      5,
    ],
    "cleavedBonds": [
      {
        "atom1": 5,
        "atom2": 6,
        "index": 5,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
      "gNp@DiYjj@xPhIP^HrXhv_hdCaUQpj\`",
      "dmT@@DiYUjjj@@xPJBJCsAdXdYaYV_hb@xUTNET",
      "do|@@DiYUft]jjhJBQJNHCa@hHhOJFQbQfEeYtCdpl_hb@xUTNET",
    ],
    "idCode": "did@\`@VTeYWaj@@@@",
    "mfInfo": {
      "atoms": {
        "C": 8,
        "H": 9,
      },
      "charge": 0,
      "mass": 105.15735396038399,
      "mf": "C8H9",
      "monoisotopicMass": 105.07042529007,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      6,
    ],
    "cleavedBonds": [
      {
        "atom1": 6,
        "atom2": 7,
        "index": 6,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gJP@DjZhCaB\`eAyBIgzI@xUT\\Jh",
      "dax@@DjUZjh@xPJBJCuADXfYaY~bHCaUPxUP",
      "dmt@@DjUVvFjjjdBbb@xPJBJCpaDXfYaYTMgY~bHCaUPxUP",
      "do|@@DjUVuYajjjjBpJJHD\`NDB\`b\`|hQFIfXVUCYvCDy~bHCaUPxUP",
    ],
    "idCode": "deT@\`@VTeVUFZh@@@@",
    "mfInfo": {
      "atoms": {
        "C": 9,
        "H": 11,
      },
      "charge": 0,
      "mass": 119.18397136523078,
      "mf": "C9H11",
      "monoisotopicMass": 119.08607535453,
      "unsaturation": 4.5,
    },
  },
  {
    "atomMap": [
      7,
    ],
    "cleavedBonds": [
      {
        "atom1": 6,
        "atom2": 7,
        "index": 6,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gJP@DjZhCaB\`eAyBIgzI@xUT\\Jh",
      "dax@@DjUZjh@xPJBJCuADXfYaY~bHCaUPxUP",
      "dmt@@DjUVvFjjjdBbb@xPJBJCpaDXfYaYTMgY~bHCaUPxUP",
      "do|@@DjUVuYajjjjBpJJHD\`NDB\`b\`|hQFIfXVUCYvCDy~bHCaUPxUP",
    ],
    "idCode": "dmL@\`@VTfUmZZUjjjj@@",
    "mfInfo": {
      "atoms": {
        "C": 10,
        "H": 17,
      },
      "charge": 0,
      "mass": 137.2423517863007,
      "mf": "C10H17",
      "monoisotopicMass": 137.13302554791,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      6,
    ],
    "cleavedBonds": [
      {
        "atom1": 5,
        "atom2": 6,
        "index": 5,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
      "gNp@DiYjj@xPhIP^HrXhv_hdCaUQpj\`",
      "dmT@@DiYUjjj@@xPJBJCsAdXdYaYV_hb@xUTNET",
      "do|@@DiYUft]jjhJBQJNHCa@hHhOJFQbQfEeYtCdpl_hb@xUTNET",
    ],
    "idCode": "dcl@\`@VTee[UnWZjjjh@@",
    "mfInfo": {
      "atoms": {
        "C": 11,
        "H": 19,
      },
      "charge": 0,
      "mass": 151.2689691911475,
      "mf": "C11H19",
      "monoisotopicMass": 151.14867561237,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      5,
    ],
    "cleavedBonds": [
      {
        "atom1": 4,
        "atom2": 5,
        "index": 4,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gC\`@Dij@xPhIP^PrY~bPNEUGBj@",
      "gNp@DiYjB@xPhIP^prYlQHGBjcaU@",
      "dmT@@DiYUj\`\`h@xPJBJCsAdYfEEPt_hb@xUTNET",
      "dg\\@@DiYUgQvhHJJ@Ca@hHhOJFQfXTUCQtCDxn_hb@xUTNET",
    ],
    "idCode": "dk\\@\`@VTeYVux]Njjjj\`@",
    "mfInfo": {
      "atoms": {
        "C": 12,
        "H": 21,
      },
      "charge": 0,
      "mass": 165.2955865959943,
      "mf": "C12H21",
      "monoisotopicMass": 165.16432567683,
      "unsaturation": 2.5,
    },
  },
  {
    "atomMap": [
      4,
    ],
    "cleavedBonds": [
      {
        "atom1": 3,
        "atom2": 4,
        "index": 3,
        "order": 1,
      },
    ],
    "fragmentType": "acyclic",
    "hoses": [
      "eF@Hp\\IPUA~b\`xUTxUP",
      "gJP@DjXHCaB\`eAzSOtRApjhxUP",
      "dax@@DjUXHH@xPJBJCqQbQdEgzH\`NEUCaU@",
      "deT@@DjUWXXHH\`@xPJBJCwQbQdEePt]gzH\`NEUCaU@",
      "dcL@@DjUWya\`\`bh@NDB\`b\`}LXdYAYTMGXLSGzH\`NEUCaU@",
    ],
    "idCode": "dg|@\`@VTeVUmQd{jjjjj@@",
    "mfInfo": {
      "atoms": {
        "C": 13,
        "H": 23,
      },
      "charge": 0,
      "mass": 179.3222040008411,
      "mf": "C13H23",
      "monoisotopicMass": 179.17997574129,
      "unsaturation": 2.5,
    },
  },
  {
    "fragmentType": "molecular ion",
    "idCode": "flu@@@DjU_YVuirshiBBJjjjJ@@",
    "mfInfo": {
      "atoms": {
        "C": 19,
        "H": 28,
      },
      "charge": 0,
      "mass": 256.4263231515315,
      "mf": "C19H28",
      "monoisotopicMass": 256.21910090244,
      "unsaturation": 6,
    },
  },
]
`;
